core/src/Cuda/Kokkos_Cuda_Instance.cpp

```diff
@@ -610,11 +605,17 @@ void CudaInternal::finalize() {
     using RecordHost =
         Kokkos::Impl::SharedAllocationRecord<CudaHostPinnedSpace>;
 
+    // FIXME multi-GPU support. We need to have a way to figure out the correct
+    // device id for allocations (which we know for the internal ones here
+    // anyway).
+    int default_device = Impl::CudaInternal::singleton().m_cudaDev;
+    Impl::CudaInternal::singleton().m_cudaDev = m_cudaDev;
     RecordCuda::decrement(RecordCuda::get_record(m_scratchFlags));
     RecordCuda::decrement(RecordCuda::get_record(m_scratchSpace));
     RecordHost::decrement(RecordHost::get_record(m_scratchUnified));
     if (m_scratchFunctorSize > 0)
       RecordCuda::decrement(RecordCuda::get_record(m_scratchFunctor));
+    Impl::CudaInternal::singleton().m_cudaDev = default_device;
```

Comment:
Where is that default instance device ID used?
Presumably the retrieving the allocation record and decrementing the reference counter do not care about device ids.  Is that when the ref counter goes to zero and we call the custom deleter encapsulating a cudaFree or whatnot?

Reply:
We are failing in https://github.com/kokkos/kokkos/blob/96d530a24034467db26070081b62a0a23c60530e/core/src/impl/Kokkos_SharedAlloc_timpl.hpp#L236-L243.
Full backtrace
```
#0  0x00007ffff64dd000 in cudbgReportDriverApiError () from /usr/lib/x86_64-linux-gnu/libcuda.so.1
#1  0x00007ffff67981eb in ?? () from /usr/lib/x86_64-linux-gnu/libcuda.so.1
#2  0x00007ffff49374e7 in ?? () from /usr/lib/x86_64-linux-gnu/libcudadebugger.so.1
#3  0x00007ffff4914c4e in ?? () from /usr/lib/x86_64-linux-gnu/libcudadebugger.so.1
#4  0x00007ffff4925bfa in ?? () from /usr/lib/x86_64-linux-gnu/libcudadebugger.so.1
#5  0x00007ffff490b2d7 in ?? () from /usr/lib/x86_64-linux-gnu/libcudadebugger.so.1
#6  0x00007ffff4a82996 in ?? () from /usr/lib/x86_64-linux-gnu/libcudadebugger.so.1
#7  0x00007ffff67b08d6 in ?? () from /usr/lib/x86_64-linux-gnu/libcuda.so.1
#8  0x00007ffff6067844 in cudaMemcpyAsync () from /usr/local/cuda/lib64/libcudart.so.11.0
#9  0x0000555555620aba in Kokkos::Impl::CudaInternal::cuda_memcpy_async_wrapper<true> (this=0x5555556ff0c0 <Kokkos::Impl::CudaInternal::singleton()::self>, dst=0x7fffffffd300, src=0x2bc0000000, count=128, 
    kind=cudaMemcpyDefault, stream=0x0) at /app/kokkos/core/src/Cuda/Kokkos_Cuda_Instance.hpp:408
#10 0x000055555561c478 in Kokkos::Impl::DeepCopyAsyncCuda (instance=..., dst=0x7fffffffd300, src=0x2bc0000000, n=128) at /app/kokkos/core/src/Cuda/Kokkos_CudaSpace.cpp:76
#11 0x0000555555621388 in Kokkos::Impl::DeepCopy<Kokkos::HostSpace, Kokkos::CudaSpace, Kokkos::Cuda, void>::DeepCopy (this=0x7fffffffd23f, instance=..., dst=0x7fffffffd300, src=0x2bc0000000, n=128)
    at /app/kokkos/core/src/Cuda/Kokkos_CudaSpace.hpp:455
#12 0x000055555561f9e7 in Kokkos::Impl::HostInaccessibleSharedAllocationRecordCommon<Kokkos::CudaSpace>::get_record (alloc_ptr=0x2bc0000080) at /app/kokkos/core/src/impl/Kokkos_SharedAlloc_timpl.hpp:238
#13 0x00005555555fca0a in Kokkos::Impl::CudaInternal::finalize (this=0x555555a0d7f0) at /app/kokkos/core/src/Cuda/Kokkos_Cuda_Instance.cpp:611
#14 0x00005555555fd4a3 in operator() (__closure=0x555557101dc0, ptr=0x555555a0d7f0) at /app/kokkos/core/src/Cuda/Kokkos_Cuda_Instance.cpp:808
#15 0x00005555555fe8fc in std::__invoke_impl<void, Kokkos::Cuda::Cuda(cudaStream_t, Kokkos::Impl::ManageStream)::<lambda(Kokkos::Impl::CudaInternal*)>&, Kokkos::Impl::CudaInternal*>(std::__invoke_other, struct {...} &) (__f=...) at /usr/include/c++/11/bits/invoke.h:61
#16 0x00005555555fe733 in std::__invoke_r<void, Kokkos::Cuda::Cuda(cudaStream_t, Kokkos::Impl::ManageStream)::<lambda(Kokkos::Impl::CudaInternal*)>&, Kokkos::Impl::CudaInternal*>(struct {...} &) (__fn=...)
    at /usr/include/c++/11/bits/invoke.h:111
#17 0x00005555555fe4b4 in std::_Function_handler<void(Kokkos::Impl::CudaInternal*), Kokkos::Cuda::Cuda(cudaStream_t, Kokkos::Impl::ManageStream)::<lambda(Kokkos::Impl::CudaInternal*)> >::_M_invoke(const std::_Any_data &, Kokkos::Impl::CudaInternal *&&) (__functor=..., __args#0=@0x7fffffffd4c0: 0x555555a0d7f0) at /usr/include/c++/11/bits/std_function.h:290
#18 0x00005555555662bf in std::function<void (Kokkos::Impl::CudaInternal*)>::operator()(Kokkos::Impl::CudaInternal*) const (this=0x555557101dc0, __args#0=0x555555a0d7f0)
    at /usr/include/c++/11/bits/std_function.h:590
#19 0x0000555555565e9a in Kokkos::Impl::HostSharedPtr<Kokkos::Impl::CudaInternal>::cleanup (this=0x7fffffffd5c0) at /app/kokkos/core/src/impl/Kokkos_HostSharedPtr.hpp:120
#20 0x0000555555565434 in Kokkos::Impl::HostSharedPtr<Kokkos::Impl::CudaInternal>::~HostSharedPtr (this=0x7fffffffd5c0, __in_chrg=<optimized out>) at /app/kokkos/core/src/impl/Kokkos_HostSharedPtr.hpp:92
#21 0x0000555555564c72 in Kokkos::Cuda::~Cuda (this=0x7fffffffd5c0, __in_chrg=<optimized out>) at /app/kokkos/core/src/Cuda/Kokkos_Cuda.hpp:97
#22 0x0000555555563ed3 in (anonymous namespace)::cuda_multi_gpu_Test::TestBody (this=0x555555776ed0) at /app/kokkos/core/unit_test/cuda/TestCuda_InterOp_StreamsMultiGPU.cpp:40
#23 0x00005555555a43d1 in testing::internal::HandleSehExceptionsInMethodIfSupported<testing::Test, void> (object=0x555555776ed0, method=&virtual testing::Test::TestBody(), location=0x555555623c4b "the test body")
    at /app/kokkos/tpls/gtest/gtest/gtest-all.cc:4101
#24 0x000055555559c237 in testing::internal::HandleExceptionsInMethodIfSupported<testing::Test, void> (object=0x555555776ed0, method=&virtual testing::Test::TestBody(), location=0x555555623c4b "the test body")
    at /app/kokkos/tpls/gtest/gtest/gtest-all.cc:4137
#25 0x000055555556ff94 in testing::Test::Run (this=0x555555776ed0) at /app/kokkos/tpls/gtest/gtest/gtest-all.cc:4176
#26 0x0000555555570a22 in testing::TestInfo::Run (this=0x555555772680) at /app/kokkos/tpls/gtest/gtest/gtest-all.cc:4326
#27 0x0000555555571329 in testing::TestSuite::Run (this=0x555555772ba0) at /app/kokkos/tpls/gtest/gtest/gtest-all.cc:4480
#28 0x000055555558122f in testing::internal::UnitTestImpl::RunAllTests (this=0x5555557727f0) at /app/kokkos/tpls/gtest/gtest/gtest-all.cc:7320
#29 0x00005555555a58c3 in testing::internal::HandleSehExceptionsInMethodIfSupported<testing::internal::UnitTestImpl, bool> (object=0x5555557727f0, 
    method=(bool (testing::internal::UnitTestImpl::*)(testing::internal::UnitTestImpl * const)) 0x555555580e3a <testing::internal::UnitTestImpl::RunAllTests()>, 
    location=0x5555556246b0 "auxiliary test code (environments or event listeners)") at /app/kokkos/tpls/gtest/gtest/gtest-all.cc:4101
#30 0x000055555559d485 in testing::internal::HandleExceptionsInMethodIfSupported<testing::internal::UnitTestImpl, bool> (object=0x5555557727f0, 
    method=(bool (testing::internal::UnitTestImpl::*)(testing::internal::UnitTestImpl * const)) 0x555555580e3a <testing::internal::UnitTestImpl::RunAllTests()>, 
    location=0x5555556246b0 "auxiliary test code (environments or event listeners)") at /app/kokkos/tpls/gtest/gtest/gtest-all.cc:4137
#31 0x000055555557f997 in testing::UnitTest::Run (this=0x5555556e1c00 <testing::UnitTest::GetInstance()::instance>) at /app/kokkos/tpls/gtest/gtest/gtest-all.cc:6903
#32 0x0000555555563af7 in RUN_ALL_TESTS () at /app/kokkos/tpls/gtest/gtest/gtest.h:12372
#33 0x0000555555563988 in main (argc=1, argv=0x7fffffffdc18) at /app/kokkos/core/unit_test/UnitTestMain.cpp:22
```

Reply:
The problem is the `memcpy` that tries to access an allocation on the wrong (default) device. The `fence` itself is fine (but also doesn't use a stream on the correct device) and wouldn't crash.

Reply:
My proposition is to first make everything work with unmanaged `Views` and have a better look at `SharedAllocationRecord` for managed `Views` separately (possibly later).